server/planning/planning/planning_history_async_service.py

```diff
@@ -0,0 +1,207 @@
+            diff = await self._get_coverage_diff(cov, original_coverage)
+            if len(diff.keys()) > 1:
+                await self._save_history(item, diff, "coverage_edited")
+
+            if original_coverage is not None:
+                if (
+                    cov.get("workflow_status") == WORKFLOW_STATE.CANCELLED
+                    and original_coverage.get("workflow_status") != WORKFLOW_STATE.CANCELLED
+                ):
+                    operation = "coverage_cancelled"
+                    diff = {
+                        "coverage_id": cov.get("coverage_id"),
+                        "workflow_status": cov["workflow_status"],
+                    }
+                    if not original.get(LOCK_ACTION):
+                        operation = "events_cancel"
+                    elif (
+                        original.get(LOCK_ACTION) == ITEM_ACTIONS.PLANNING_CANCEL  # type: ignore
+                        or updates.get("state") == WORKFLOW_STATE.CANCELLED
+                    ):
+                        # If cancelled through item action or through editor
+                        operation = "planning_cancel"
+
+                    await self._save_history(item, diff, operation)
+
+                # If assignment was added in an update
+                if cov.get("assigned_to", {}).get("assignment_id") and not (
+                    original_coverage.get("assigned_to") or {}
+                ).get("assignment_id"):
+                    diff = {
+                        "coverage_id": cov.get("coverage_id"),
+                        "assigned_to": cov["assigned_to"],
+                    }
+                    await self._save_history(item, diff, "coverage_assigned")
+
+        for cov in deleted:
+            await self._save_history(item, {"coverage_id": cov.get("coverage_id")}, "coverage_deleted")
+
+    async def on_spike(self, updates: dict[str, Any], original: dict[str, Any]):
+        """Spike event
+
+        On spike of a planning item the history of any agendas that the item belongs to will have an entry added to
+        their history as effectively the scope of what the agenda contains has changed.
+        :param updates:
+        :param original:
+        :return:
+        """
+        await super().on_spike(updates, original)
+
+    async def on_unspike(self, updates: dict[str, Any], original: dict[str, Any]):
+        await super().on_unspike(updates, original)
+
+    async def on_duplicate(self, parent: dict[str, Any], duplicate: dict[str, Any]):
+        await self._save_history(
+            {ID_FIELD: str(parent[ID_FIELD])},
+            {"duplicate_id": str(duplicate[ID_FIELD])},
+            "duplicate",
+        )
+
+    async def on_duplicate_from(self, item: dict[str, Any], duplicate_id: str):
+        new_plan = deepcopy(item)
+        new_plan["duplicate_id"] = duplicate_id
+        await self._save_history({ID_FIELD: str(item[ID_FIELD])}, new_plan, "duplicate_from")
```

Comment:
It would be good to add some docstrings to at least some of these methods to clarify their purposes.